lib/extend/injector.js

```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+class Injector {
+  constructor() {
+    this.store = {
+      head_begin: new Set(),
+      head_end: new Set(),
+      body_begin: new Set(),
+      body_end: new Set()
+    };
+  }
+
+  list() {
+    return this.store;
+  }
+
+  get(entry) {
+    return Array.from(this.store[entry]);
+  }
+
+  register(entry, value, to = '') {
+    if (!entry) throw new TypeError('entry is required');
+    if (typeof value === 'function') value = value();
+
+    this.store[entry].add({ value, to });
```

Comment:
https://github.com/hexojs/hexo/issues/4047#issuecomment-570886958

However, `locals` is only available in `after_route_render` filter. Will be implement after #4051 is merged.